pkg/process/metadata/parser/service.go

```diff
@@ -0,0 +1,200 @@
+package parser
+
+import (
+	"path/filepath"
+	"strings"
+	"unicode"
+
+	"github.com/DataDog/datadog-agent/pkg/process/metadata"
+	"github.com/DataDog/datadog-agent/pkg/process/procutil"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+type serviceExtractorFn func(args []string) string
+
+// List of binaries that usually have additional process context of whats running
+var binsWithContext = map[string]serviceExtractorFn{
+	"python":    parseCommandContextPython,
+	"python2.7": parseCommandContextPython,
+	"python3":   parseCommandContextPython,
+	"python3.7": parseCommandContextPython,
+	"ruby2.3":   parseCommandContext,
+	"ruby":      parseCommandContext,
+	"java":      parseCommandContext,
+	"sudo":      parseCommandContext,
+}
+
+var _ metadata.Extractor = &ServiceExtractor{}
+
+// ServiceExtractor infers a service tag by extracting it from a process
+type ServiceExtractor struct {
+	Enabled      bool
+	serviceByPID map[int32]*serviceMetadata
+}
+
+type serviceMetadata struct {
+	cmdline    []string
+	serviceTag string
+}
```

Comment:
I don't think I see `cmdline` being used anywhere so it seems that `serviceByPID` could just be a plain `map[int32]string`, no?

Reply:
The `Extract` function compares the current commandline with the one retrieved from `serviceByPid` to avoid parsing an already known process.